codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrame.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+/**
+ * A frame whose meaning <em>may</em> apply to a particular stream, instead of the entire
+ * connection. It is still possibly for this frame type to apply to the entire connection. In such
+ * cases, the {@code stream} reference should be {@code null} or an object referring to the
+ * connection.
+ *
+ * <p>The meaning of {@code stream} is context-dependent and may change as a frame is processed in
+ * the pipeline.
+ */
+public interface Http2StreamFrame extends Http2Frame {
+    /**
+     * Set the stream identifier for this message.
+     *
+     * @return {@code this}
+     */
+    Http2StreamFrame setStream(Object stream);
```

Comment:
> We are really setting the associated channel right? Should this be channel(...)?

Right now my plan is to change the value over time for a particular message. The actual meaning of the value is an internal detail and users of it would at-worst seem to only copy from an incoming message to an outgoing message. In a future PR, the value stored here will vary between the child-channel and an Http2Stream, depending on where in the pipeline it is.

Note that while on a child channel, the value of the stream is required to be null (both sending and receiving). It is only used briefly when being passed from the child channel to the parent channel; i.e., no application code could actually see the value stored as the `stream` today; it is internal bookkeeping.

This is part of a larger design. I'm going to be splitting the child-channel piece from "turn frames into objects" piece. In the future, the most important value stored here will be the Http2Stream. Once I do the split, I think we'd be in a much better position to discuss alternative designs for storing this value.

> I wonder if we could assume a different construction pattern where basically the Http2MultiplexCodec (or some class) is a factory for all HttpFrame objects.

In the case of having a factory for `HttpFrame` objects, that seems to completely prevent any extending/custom implementations. I thought you were pushing for being able to support custom frame types and the like?

The only way I see communicate such a factory would be a channel attribute, which seems weak.

A factory solution seems much more restrictive than the problem.

> This way we may have better control of the lifetime of "outbound" created stream, and also the channel can be final in all Http2StreamFrame classes. 

We already have a clear idea of the lifetime of the created stream. But even so, a factory wouldn't tell us anything about lifetime, because objects can be created long before they reach the head of the pipeline.

Making the channel final would be counter-productive for my objectives. I'm trying to provide a mechanism that would allow unknown HTTP/2 frame type pass-through and prevent allocations/copies.